File: backend/internal/weather/outlook.go
package weather

import (
	"context"
	"fmt"
	"math"
	"sort"
	"sync"
	"time"
)

type OutlookModelDay struct {
	Date                     string   `json:"date"`
	TemperatureMin           float64  `json:"temperatureMin"`
	TemperatureMax           float64  `json:"temperatureMax"`
	PrecipitationProbability *float64 `json:"precipitationProbability"`
	Precipitation            float64  `json:"precipitation"`
}

type OutlookModel struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Short       string            `json:"short"`
	HorizonDays int               `json:"horizonDays"`
	Daily       []OutlookModelDay `json:"daily"`
}

type EnsembleDay struct {
	Date                 string    `json:"date"`
	TemperatureMedian    float64   `json:"temperatureMedian"`
	TemperatureP10       float64   `json:"temperatureP10"`
	TemperatureP90       float64   `json:"temperatureP90"`
	PrecipitationMedian  float64   `json:"precipitationMedian"`
	PrecipitationP10     float64   `json:"precipitationP10"`
	PrecipitationP90     float64   `json:"precipitationP90"`
	TemperatureMembers   []float64 `json:"-"`
	PrecipitationMembers []float64 `json:"-"`
}

type EnsembleModel struct {
	ID                string        `json:"id"`
	Name              string        `json:"name"`
	Short             string        `json:"short"`
	MemberCount       int           `json:"memberCount"`
	Daily             []EnsembleDay `json:"daily"`
	ExcludeFromFusion bool          `json:"-"`
}

type FusionDay struct {
	Date                string  `json:"date"`
	TemperatureP10      float64 `json:"temperatureP10"`
	TemperatureP25      float64 `json:"temperatureP25"`
	TemperatureP50      float64 `json:"temperatureP50"`
	TemperatureP75      float64 `json:"temperatureP75"`
	TemperatureP90      float64 `json:"temperatureP90"`
	PrecipitationP10    float64 `json:"precipitationP10"`
	PrecipitationP50    float64 `json:"precipitationP50"`
	PrecipitationP90    float64 `json:"precipitationP90"`
	RainProbability1mm  float64 `json:"rainProbability1mm"`
	RainProbability10mm float64 `json:"rainProbability10mm"`
	ModelCount          int     `json:"modelCount"`
	MemberCount         int     `json:"memberCount"`
}

type Fusion struct {
	Method string      `json:"method"`
	Daily  []FusionDay `json:"daily"`
	Notice string      `json:"notice"`
}

type Outlook struct {
	Mode        string          `json:"mode"`
	HorizonDays int             `json:"horizonDays"`
	Models      []OutlookModel  `json:"models,omitempty"`
	Ensembles   []EnsembleModel `json:"ensembles,omitempty"`
	Fusion      *Fusion         `json:"fusion,omitempty"`
	Notice      string          `json:"notice"`
	Warnings    []string        `json:"warnings,omitempty"`
	RefreshedAt time.Time       `json:"refreshedAt"`
	Source      string          `json:"source"`
}

type OutlookProvider interface {
	ModelOutlook(context.Context, Coordinates) ([]OutlookModel, []string, error)
	EnsembleOutlook(context.Context, Coordinates) ([]EnsembleModel, []string, error)
}

type outlookCacheEntry struct {
	value     Outlook
	expiresAt time.Time
}

type OutlookService struct {
	provider OutlookProvider
	now      func() time.Time
	mu       sync.RWMutex
	cache    map[string]outlookCacheEntry
}

func NewOutlookService(provider OutlookProvider) *OutlookService {
	return &OutlookService{
		provider: provider,
		now:      time.Now,
		cache:    make(map[string]outlookCacheEntry),
	}
}

func (s *OutlookService) Get(ctx context.Context, coordinates Coordinates, view string) (Outlook, error) {
	if view != "16" && view != "30" {
		return Outlook{}, fmt.Errorf("unsupported outlook view %q", view)
	}
	key := fmt.Sprintf("%.5f:%.5f:%s", coordinates.Latitude, coordinates.Longitude, view)
	s.mu.RLock()
	cached, ok := s.cache[key]
	s.mu.RUnlock()
	if ok && s.now().Before(cached.expiresAt) {
		cached.value.Source = "cache"
		return cached.value, nil
	}

	var result Outlook
	var err error
	if view == "16" {
		result.Models, result.Warnings, err = s.provider.ModelOutlook(ctx, coordinates)
		result.Mode = "models"
		result.HorizonDays = 16
		result.Notice = "Die Linien enden an der nativen Reichweite des jeweiligen Modells. Ab Tag 8 sinkt die räumliche Präzision deutlich."
	} else {
		result.Ensembles, result.Warnings, err = s.provider.EnsembleOutlook(ctx, coordinates)
		result.Mode = "ensemble"
		result.HorizonDays = 30
		if err == nil {
			result.Fusion = fuseEnsembles(result.Ensembles)
		}
		result.Notice = "Das Band zeigt P10 bis P90 der Ensembleläufe. Es ist ein Wahrscheinlichkeitsraum, keine garantierte Tagesprognose."
	}
	if err != nil {
		if ok {
			cached.value.Source = "stale"
			return cached.value, nil
		}
		return Outlook{}, err
	}
	result.RefreshedAt = s.now().UTC()
	result.Source = "refresh"

	s.mu.Lock()
	s.cache[key] = outlookCacheEntry{value: result, expiresAt: s.now().Add(30 * time.Minute)}
	s.mu.Unlock()
	return result, nil
}

type fusionMemberSet struct {
	temperature   []float64
	precipitation []float64
}

type weightedValue struct {
	value  float64
	weight float64
}

func fuseEnsembles(models []EnsembleModel) *Fusion {
	byDate := make(map[string][]fusionMemberSet)
	for _, model := range models {
		if model.ExcludeFromFusion {
			continue
		}
		for _, day := range model.Daily {
			temperatures, precipitation := pairedFiniteMembers(day.TemperatureMembers, day.PrecipitationMembers)
			if len(temperatures) == 0 {
				continue
			}
			byDate[day.Date] = append(byDate[day.Date], fusionMemberSet{
				temperature: temperatures, precipitation: precipitation,
			})
		}
	}
	if len(byDate) == 0 {
		return nil
	}

	dates := make([]string, 0, len(byDate))
	for date := range byDate {
		dates = append(dates, date)
	}
	sort.Strings(dates)

	daily := make([]FusionDay, 0, len(dates))
	for _, date := range dates {
		memberSets := byDate[date]
		temperatures, precipitation := balancedSamples(memberSets)
		memberCount := 0
		for _, members := range memberSets {
			memberCount += len(members.temperature)
		}
		daily = append(daily, FusionDay{
			Date:                date,
			TemperatureP10:      roundFusion(weightedQuantile(temperatures, .10), 1),
			TemperatureP25:      roundFusion(weightedQuantile(temperatures, .25), 1),
			TemperatureP50:      roundFusion(weightedQuantile(temperatures, .50), 1),
			TemperatureP75:      roundFusion(weightedQuantile(temperatures, .75), 1),
			TemperatureP90:      roundFusion(weightedQuantile(temperatures, .90), 1),
			PrecipitationP10:    roundFusion(weightedQuantile(precipitation, .10), 2),
			PrecipitationP50:    roundFusion(weightedQuantile(precipitation, .50), 2),
			PrecipitationP90:    roundFusion(weightedQuantile(precipitation, .90), 2),
			RainProbability1mm:  roundFusion(weightedProbabilityAtLeast(precipitation, 1)*100, 1),
			RainProbability10mm: roundFusion(weightedProbabilityAtLeast(precipitation, 10)*100, 1),
			ModelCount:          len(memberSets),
			MemberCount:         memberCount,
		})
	}

	return &Fusion{
		Method: "equal-model-weighted-empirical",
		Daily:  daily,
		Notice: "Rohfusion ohne historische Kalibrierung: Pro Tag erh\u00e4lt jedes verf\u00fcgbare Kurz- oder Mittelfristmodell dasselbe Gesamtgewicht; dessen Mitglieder teilen es gleichm\u00e4\u00dfig. EC46 bleibt als separate erweiterte Orientierung au\u00dferhalb der Tagesfusion.",
	}
}

func pairedFiniteMembers(temperatures, precipitation []float64) ([]float64, []float64) {
	count := len(temperatures)
	if len(precipitation) < count {
		count = len(precipitation)
	}
	cleanTemperatures := make([]float64, 0, count)
	cleanPrecipitation := make([]float64, 0, count)
	for index := 0; index < count; index++ {
		if !finite(temperatures[index]) || !finite(precipitation[index]) {
			continue
		}
		cleanTemperatures = append(cleanTemperatures, temperatures[index])
		cleanPrecipitation = append(cleanPrecipitation, precipitation[index])
	}
	return cleanTemperatures, cleanPrecipitation
}

func balancedSamples(models []fusionMemberSet) ([]weightedValue, []weightedValue) {
	if len(models) == 0 {
		return nil, nil
	}
	temperatures := make([]weightedValue, 0)
	precipitation := make([]weightedValue, 0)
	modelWeight := 1 / float64(len(models))
	for _, model := range models {
		if len(model.temperature) == 0 {
			continue
		}
		memberWeight := modelWeight / float64(len(model.temperature))
		for index := range model.temperature {
			temperatures = append(temperatures, weightedValue{value: model.temperature[index], weight: memberWeight})
			precipitation = append(precipitation, weightedValue{value: model.precipitation[index], weight: memberWeight})
		}
	}
	return temperatures, precipitation
}

func weightedQuantile(values []weightedValue, probability float64) float64 {
	ordered := make([]weightedValue, 0, len(values))
	for _, item := range values {
		if validWeightedValue(item) {
			ordered = append(ordered, item)
		}
	}
	if len(ordered) == 0 {
		return 0
	}
	sort.SliceStable(ordered, func(left, right int) bool { return ordered[left].value < ordered[right].value })
	totalWeight := 0.0
	for _, item := range ordered {
		totalWeight += item.weight
	}
	probability = math.Max(0, math.Min(1, probability))
	target := probability * totalWeight
	cumulative := 0.0
	for _, item := range ordered {
		cumulative += item.weight
		if cumulative >= target {
			return item.value
		}
	}
	return ordered[len(ordered)-1].value
}

func weightedProbabilityAtLeast(values []weightedValue, threshold float64) float64 {
	totalWeight, eventWeight := 0.0, 0.0
	for _, item := range values {
		if !validWeightedValue(item) {
			continue
		}
		totalWeight += item.weight
		if item.value >= threshold {
			eventWeight += item.weight
		}
	}
	if totalWeight == 0 {
		return 0
	}
	return eventWeight / totalWeight
}

func validWeightedValue(item weightedValue) bool {
	return item.weight > 0 && finite(item.weight) && finite(item.value)
}

func finite(value float64) bool {
	return !math.IsNaN(value) && !math.IsInf(value, 0)
}

func roundFusion(value float64, precision int) float64 {
	factor := math.Pow10(precision)
	return math.Round(value*factor) / factor
}

File: frontend/src/adapters/browser-weather/outlook.ts
import type {
  EnsembleDay,
  EnsembleModel,
  FusionDay,
  Outlook,
  OutlookFusion,
  OutlookModel,
  OutlookView,
} from '../../types/outlook'
import {
  cached,
  ENSEMBLE_URL,
  fetchJSON,
  FORECAST_URL,
  quantile,
  queryURL,
  round1,
  SEASONAL_URL,
} from './shared'

type NullableSeries = Array<number | null>
type DailyPayload = Record<string, string[] | NullableSeries>

interface DynamicResponse {
  daily: DailyPayload
}

async function fetchDaily(
  endpoint: string,
  values: Record<string, string | number>,
  signal?: AbortSignal,
) {
  const response = await fetchJSON<DynamicResponse>(queryURL(endpoint, values), signal)
  if (!response.daily) throw new Error('Open-Meteo liefert keine Tagesdaten.')
  return response.daily
}

function numberSeries(daily: DailyPayload, key: string): NullableSeries {
  const values = daily[key]
  return Array.isArray(values) ? values as NullableSeries : []
}

function dates(daily: DailyPayload) {
  return daily.time as string[] ?? []
}

async function modelOutlook(
  latitude: number,
  longitude: number,
  signal?: AbortSignal,
): Promise<Outlook> {
  const daily = await fetchDaily(FORECAST_URL, {
    latitude: latitude.toFixed(5),
    longitude: longitude.toFixed(5),
    models: 'icon_seamless,ecmwf_ifs025,ecmwf_aifs025_single,gfs_seamless',
    daily: 'temperature_2m_max,temperature_2m_min,precipitation_probability_max,precipitation_sum',
    forecast_days: 16,
    timezone: 'auto',
  }, signal)
  const definitions = [
    { id: 'icon', name: 'DWD ICON', short: 'ICON', suffix: 'icon_seamless' },
    { id: 'ifs', name: 'ECMWF IFS', short: 'IFS', suffix: 'ecmwf_ifs025' },
    { id: 'aifs', name: 'ECMWF AIFS', short: 'AIFS', suffix: 'ecmwf_aifs025_single' },
    { id: 'gfs', name: 'NOAA GFS', short: 'GFS', suffix: 'gfs_seamless' },
  ]
  const warnings: string[] = []
  const models = definitions.flatMap<OutlookModel>((definition) => {
    const minimums = numberSeries(daily, `temperature_2m_min_${definition.suffix}`)
    const maximums = numberSeries(daily, `temperature_2m_max_${definition.suffix}`)
    if (!minimums.length || !maximums.length) {
      warnings.push(`${definition.short} nicht verfügbar`)
      return []
    }
    const probabilities = numberSeries(daily, `precipitation_probability_max_${definition.suffix}`)
    const precipitation = numberSeries(daily, `precipitation_sum_${definition.suffix}`)
    const points = dates(daily).flatMap((date, index) => {
      const temperatureMin = minimums[index]
      const temperatureMax = maximums[index]
      if (temperatureMin === null || temperatureMax === null || temperatureMin === undefined || temperatureMax === undefined) return []
      return [{
        date,
        temperatureMin,
        temperatureMax,
        precipitationProbability: probabilities[index] ?? null,
        precipitation: precipitation[index] ?? 0,
      }]
    })
    return [{
      id: definition.id,
      name: definition.name,
      short: definition.short,
      horizonDays: points.length,
      daily: points,
    }]
  })
  if (models.length < 2) throw new Error('Zu wenige Langfristmodelle sind erreichbar.')
  return {
    mode: 'models',
    horizonDays: 16,
    models,
    notice: 'Die Linien enden an der nativen Reichweite des jeweiligen Modells. Ab Tag 8 sinkt die räumliche Präzision deutlich.',
    warnings,
    refreshedAt: new Date().toISOString(),
    source: 'refresh',
  }
}

function memberSeries(daily: DailyPayload, prefix: string) {
  return Object.keys(daily)
    .filter((key) => key === prefix || key.startsWith(`${prefix}_member`))
    .sort()
    .map((key) => numberSeries(daily, key))
}

function valuesAt(series: NullableSeries[], index: number) {
  return series.flatMap((member) => {
    const value = member[index]
    return value === null || value === undefined ? [] : [value]
  })
}

interface EnsembleDefinition {
  endpoint: string
  model: string
  id: string
  name: string
  short: string
  forecastDays: number
  includeInFusion: boolean
}

interface EnsembleMemberDay {
  date: string
  temperature: number[]
  precipitation: number[]
}

interface EnsembleResult {
  model: EnsembleModel
  members: EnsembleMemberDay[]
  includeInFusion: boolean
}

function weightedQuantile(groups: number[][], probability: number) {
  const usable = groups.filter((group) => group.length > 0)
  if (!usable.length) return 0
  const modelWeight = 1 / usable.length
  const samples = usable.flatMap((group) => {
    const memberWeight = modelWeight / group.length
    return group.map((value) => ({ value, weight: memberWeight }))
  }).sort((a, b) => a.value - b.value)
  const target = Math.max(0, Math.min(1, probability))
  let cumulative = 0
  for (const sample of samples) {
    cumulative += sample.weight
    if (cumulative + Number.EPSILON >= target) return round1(sample.value)
  }
  return round1(samples.at(-1)?.value ?? 0)
}

function balancedProbability(groups: number[][], threshold: number) {
  const usable = groups.filter((group) => group.length > 0)
  if (!usable.length) return 0
  const probability = usable.reduce((sum, group) =>
    sum + group.filter((value) => value >= threshold).length / group.length, 0) / usable.length
  return round1(probability * 100)
}

function buildFusion(results: EnsembleResult[]): OutlookFusion {
  const eligible = results.filter((result) => result.includeInFusion)
  const allDates = [...new Set(eligible.flatMap((result) =>
    result.members.map((day) => day.date),
  ))].sort()
  const daily = allDates.flatMap<FusionDay>((date) => {
    const activeDays = eligible.flatMap((result) => {
      const day = result.members.find((candidate) => candidate.date === date)
      return day?.temperature.length ? [day] : []
    })
    if (!activeDays.length) return []
    const temperatureGroups = activeDays.map((day) => day.temperature)
    const precipitationGroups = activeDays
      .map((day) => day.precipitation)
      .filter((values) => values.length > 0)
    return [{
      date,
      temperatureP10: weightedQuantile(temperatureGroups, 0.1),
      temperatureP25: weightedQuantile(temperatureGroups, 0.25),
      temperatureP50: weightedQuantile(temperatureGroups, 0.5),
      temperatureP75: weightedQuantile(temperatureGroups, 0.75),
      temperatureP90: weightedQuantile(temperatureGroups, 0.9),
      precipitationP10: weightedQuantile(precipitationGroups, 0.1),
      precipitationP50: weightedQuantile(precipitationGroups, 0.5),
      precipitationP90: weightedQuantile(precipitationGroups, 0.9),
      rainProbability1mm: balancedProbability(precipitationGroups, 1),
      rainProbability10mm: balancedProbability(precipitationGroups, 10),
      modelCount: activeDays.length,
      memberCount: temperatureGroups.reduce((sum, values) => sum + values.length, 0),
    }]
  })
  return {
    method: 'equal-model-weighted-empirical',
    daily,
    notice: 'Jedes verfügbare Kurz- und Mittelfristmodell erhält pro Tag dasselbe Gewicht; seine Mitglieder teilen sich dieses Gewicht. Die Ereigniswerte sind rohe, modellbalancierte Ensemble-Wahrscheinlichkeiten und noch nicht historisch kalibriert. EC46 bleibt wegen seiner gröberen Skala eine separate Langfrist-Referenz.',
  }
}

async function ensembleModel(
  definition: EnsembleDefinition,
  latitude: number,
  longitude: number,
  signal?: AbortSignal,
): Promise<EnsembleResult> {
  const daily = await fetchDaily(definition.endpoint, {
    latitude: latitude.toFixed(5),
    longitude: longitude.toFixed(5),
    models: definition.model,
    daily: 'temperature_2m_mean,precipitation_sum',
    forecast_days: definition.forecastDays,
    timezone: 'auto',
  }, signal)
  const temperatures = memberSeries(daily, 'temperature_2m_mean')
  const precipitation = memberSeries(daily, 'precipitation_sum')
  if (!temperatures.length) throw new Error(`${definition.short} liefert keine Ensembleläufe.`)
  const members = dates(daily).flatMap<EnsembleMemberDay>((date, index) => {
    const temperatureValues = valuesAt(temperatures, index)
    if (!temperatureValues.length) return []
    return [{
      date,
      temperature: temperatureValues,
      precipitation: valuesAt(precipitation, index),
    }]
  })
  const points = members.map<EnsembleDay>((day) => ({
    date: day.date,
    temperatureMedian: round1(quantile(day.temperature, 0.5)),
    temperatureP10: round1(quantile(day.temperature, 0.1)),
    temperatureP90: round1(quantile(day.temperature, 0.9)),
    precipitationMedian: round1(quantile(day.precipitation, 0.5)),
    precipitationP10: round1(quantile(day.precipitation, 0.1)),
    precipitationP90: round1(quantile(day.precipitation, 0.9)),
  }))
  return {
    model: {
      id: definition.id,
      name: definition.name,
      short: definition.short,
      memberCount: temperatures.length,
      daily: points,
    },
    members,
    includeInFusion: definition.includeInFusion,
  }
}

async function ensembleOutlook(
  latitude: number,
  longitude: number,
  signal?: AbortSignal,
): Promise<Outlook> {
  const definitions: EnsembleDefinition[] = [
    { endpoint: ENSEMBLE_URL, model: 'dwd_icon_eu_eps', id: 'icon-eu', name: 'DWD ICON-EU EPS', short: 'ICON-EU', forecastDays: 5, includeInFusion: true },
    { endpoint: ENSEMBLE_URL, model: 'ecmwf_ifs025_ensemble', id: 'ifs-ens', name: 'ECMWF IFS ENS', short: 'IFS ENS', forecastDays: 15, includeInFusion: true },
    { endpoint: ENSEMBLE_URL, model: 'ecmwf_aifs025_ensemble', id: 'aifs-ens', name: 'ECMWF AIFS ENS', short: 'AIFS ENS', forecastDays: 15, includeInFusion: true },
    { endpoint: ENSEMBLE_URL, model: 'ncep_gefs05', id: 'gefs', name: 'NOAA GEFS 0.5°', short: 'GEFS', forecastDays: 30, includeInFusion: true },
    { endpoint: ENSEMBLE_URL, model: 'google_weathernext2_ensemble', id: 'weathernext2', name: 'Google WeatherNext 2', short: 'WN2', forecastDays: 15, includeInFusion: true },
    { endpoint: SEASONAL_URL, model: 'ecmwf_ec46', id: 'ec46', name: 'ECMWF EC46', short: 'EC46', forecastDays: 30, includeInFusion: false },
  ]
  const outcomes = await Promise.allSettled(definitions.map((definition) =>
    ensembleModel(
      definition,
      latitude,
      longitude,
      signal,
    ),
  ))
  const successful = outcomes.flatMap((outcome) => outcome.status === 'fulfilled' ? [outcome.value] : [])
  const warnings = outcomes.flatMap((outcome, index) =>
    outcome.status === 'rejected' ? [`${definitions[index].short} nicht verfügbar`] : [],
  )
  if (!successful.length) throw new Error('Die Ensemblemodelle sind momentan nicht erreichbar.')
  const fusion = buildFusion(successful)
  return {
    mode: 'ensemble',
    horizonDays: fusion.daily.length,
    ensembles: successful.map((result) => result.model),
    fusion,
    notice: 'P10 bis P90 markieren den modellierten Wahrscheinlichkeitsraum. Mit wachsendem Horizont sinken Modellzahl und räumliche Präzision.',
    warnings,
    refreshedAt: new Date().toISOString(),
    source: 'refresh',
  }
}

export function getBrowserOutlook(
  view: OutlookView,
  latitude: number,
  longitude: number,
  signal?: AbortSignal,
) {
  return cached(
    `outlook:v2:${view}:${latitude.toFixed(4)}:${longitude.toFixed(4)}`,
    30 * 60_000,
    () => view === '16'
      ? modelOutlook(latitude, longitude, signal)
      : ensembleOutlook(latitude, longitude, signal),
  )
}
